main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification3.scala

```diff
@@ -0,0 +1,308 @@
+    */
+  private def fromSetFormula(t: SetFormula, loc: SourceLocation)(implicit m: Bimap[Atom, Int]): Type = t match {
+    case SetFormula.Univ => Type.Univ
+    case SetFormula.Empty => Type.Pure
+    case SetFormula.Cst(c) => fromAtom(m.getBackward(c).get, loc) // Safe: We never introduce new variables.
+    case SetFormula.Var(x) => fromAtom(m.getBackward(x).get, loc) // Safe: We never introduce new variables.
+    case SetFormula.ElemSet(x) => Type.mkUnion(x.toList.map(m.getBackward(_).get).map(fromAtom(_, loc)), loc)
+    case SetFormula.Compl(t) => Type.mkComplement(fromSetFormula(t, loc), loc)
+    case SetFormula.Inter(posElem, posCsts, posVars, negElem, negCsts, negVars, rest) =>
+      val ts = posElem.iterator.map(fromSetFormula(_, loc)) ++
+        posCsts.iterator.map(fromSetFormula(_, loc)) ++
+        posVars.iterator.map(fromSetFormula(_, loc)) ++
+        negElem.iterator.map(SetFormula.mkCompl(_)).map(fromSetFormula(_, loc)) ++
+        negCsts.iterator.map(SetFormula.mkCompl(_)).map(fromSetFormula(_, loc)) ++
+        negVars.iterator.map(SetFormula.mkCompl(_)).map(fromSetFormula(_, loc)) ++
+        rest.iterator.map(fromSetFormula(_, loc))
+      Type.mkIntersection(ts.toList, loc)
+    case SetFormula.Union(posElem, posCsts, posVars, negElem, negCsts, negVars, rest) =>
+      val ts = posElem.iterator.map(fromSetFormula(_, loc)) ++
+        posCsts.iterator.map(fromSetFormula(_, loc)) ++
+        posVars.iterator.map(fromSetFormula(_, loc)) ++
+        negElem.iterator.map(SetFormula.mkCompl(_)).map(fromSetFormula(_, loc)) ++
+        negCsts.iterator.map(SetFormula.mkCompl(_)).map(fromSetFormula(_, loc)) ++
+        negVars.iterator.map(SetFormula.mkCompl(_)).map(fromSetFormula(_, loc)) ++
+        rest.iterator.map(fromSetFormula(_, loc))
+      Type.mkUnion(ts.toList, loc)
+  }
+
+  /** Returns the Type represented by the given Atom. */
+  private def fromAtom(s: Atom, loc: SourceLocation)(implicit m: Bimap[Atom, Int]): Type = s match {
+    case Atom.Eff(sym) => Type.Cst(TypeConstructor.Effect(sym), loc)
+    case Atom.Var(sym) => Type.Var(sym, loc)
+    case Atom.Assoc(sym, arg0, kind) =>
+      val arg = fromAtom(arg0, loc)
+      Type.AssocType(AssocTypeConstructor(sym, loc), arg, kind, loc)
+    case Atom.Error(id, kind) => Type.Cst(TypeConstructor.Error(id, kind), loc)
+  }
+
+  /**
+    * Atomic terms that may appear in the algebra of effects when associated types are fully reduced.
+    */
+  private sealed trait Atom
+
+  private object Atom {
+    /** An atom representing a variable. */
+    case class Var(sym: Symbol.KindedTypeVarSym) extends Atom
+
+    /** An atom representing an effect constant. */
+    case class Eff(sym: Symbol.EffectSym) extends Atom
+
+    /** An atom representing an associated effect. */
+    case class Assoc(sym: Symbol.AssocTypeSym, arg: Atom, kind: Kind) extends Atom
+
+    /** An atom representing an invalid type. */
+    case class Error(id: Int, kind: Kind) extends Atom
+  }
+
```

Comment:
@mlutze How much of the code in this file will stay after your rewrite? 

I assume at least some of it-- so that should be cleaned up?

Reply:
Pretty much the whole file will stay after rewrite.

Reply:
OK; then it makes sense to polish